steward/src/cork.rs

```diff
@@ -10,74 +13,57 @@ use abscissa_core::{
 
 pub struct CorkHandler;
 
 #[async_trait]
 impl steward::contract_call_server::ContractCall for CorkHandler {
-    async fn submit(
+    async fn schedule(
         &self,
-        request: Request<SubmitRequest>,
-    ) -> Result<Response<SubmitResponse>, Status> {
+        request: Request<ScheduleRequest>,
+    ) -> Result<Response<ScheduleResponse>, Status> {
         let request = request.get_ref().to_owned();
-
-        // Check if cellar is governance approved before building cork
-        let config = APP.config();
-        let mut client = match CorkQueryClient::connect(config.cosmos.grpc.clone()).await {
-            Ok(c) => c,
-            Err(err) => {
-                error!("cork query client connection failed: {}", err);
-                return Err(Status::new(
-                    Code::Internal,
-                    format!("failed to query chain to validate cellar id: {}", err),
-                ));
+        let cellar_id = request.cellar_id.clone();
+        if let Err(err) = cellars::validate_cellar_id(&cellar_id).await {
+            let message = format!("cellar ID validation failed: {}", err);
+            match err.kind() {
+                CacheError => return Err(Status::new(Code::Internal, message)),
+                SPCallError => return Err(Status::new(Code::InvalidArgument, message)),
+                UnapprovedCellar => return Err(Status::new(Code::InvalidArgument, message)),
+                _ => return Err(Status::new(Code::Unknown, message)),
             }
-        };
-
-        debug!("checking if cellar ID is approved");
-        let ids = &client
-            .query_cellar_i_ds(QueryCellarIDsRequest {})
-            .await?
-            .get_ref()
-            .cellar_ids
-            .clone();
-        if !ids.contains(&request.cellar_id) {
-            info!(
-                "rejecting request for unapproved cellar {}",
-                request.cellar_id
-            );
-            return Err(Status::new(
-                Code::PermissionDenied,
-                format!(
-                    "cellar ID {} not approved by governance",
-                    &request.cellar_id
-                ),
-            ));
         }
 
         // Build and send cork
         let cellar_id = request.cellar_id.clone();
-        let cork = match build_cork(request).await {
+        let height = request.block_height;
+        let encoded_call = match get_encoded_call(request) {
             Ok(c) => c,
             Err(err) => {
-                warn!("failed to build cork for cellar {}: {}", cellar_id, err);
+                warn!("failed to get encoded call for {}: {}", cellar_id, err);
                 return Err(Status::new(Code::InvalidArgument, err.to_string()));
             }
         };
-        debug!("cork: {:?}", cork);
+        debug!("cork: {:?}", encoded_call);
 
-        if let Err(err) = send_cork(cork).await {
+        if let Err(err) = schedule_cork(&cellar_id, encoded_call.clone(), height).await {
```

Comment:
Should we be applying the same retry and error handling logic we've introduced for proposals in our main cork scheduling flow?

Reply:
discussed in chat, future PR